clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -422,7 +434,17 @@ public boolean hasUnsent() {
                                         break;
                                     } else {
                                         RecordBatch batch = deque.pollFirst();
-                                        batch.close();
+                                        if (transactionState != null) {
+                                            log.debug("Dest: {} : pid: {}, epoch: {}, Assigning sequence for {}-{} : {}", node,
+                                                    transactionState.pidAndEpoch().pid, transactionState.pidAndEpoch().epoch,
+                                                    batch.topicPartition.topic(), batch.topicPartition.partition(),
+                                                    transactionState.sequenceNumber(batch.topicPartition));
```

Comment:
nit: we can call `sequenceNumber()` once and cache the value locally, since the function itself is synchronized an hence may be expensive. Similarly for `pidAndEpoch()`.